### PR TITLE
rgw: fix sharded bucket listing with prefix/delimiter

### PR DESCRIPTION
**Pull request by cbodley**
fixes some s3test failures against sharded buckets, where we didn't properly handle duplicate common prefixes from different shards

Fixes: https://tracker.ceph.com/issues/44353

### CODE REVIEW

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -8281,6 +8281,17 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
...
+    bool next_candidate(std::map<std::string, size_t>& candidates,
+                        size_t tracker_idx) {
+      while (!at_end()) {
+        if (candidates.emplace(entry_name(), tracker_idx).second) {
+          return true;
+        }
+        advance(); // skip duplicate common prefixes
+      }
+      return false;
+    }
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc]: Not sure I like the organization of this. This is logic from the outside pushed into the struct, which is realy only a tracker (a fancy iterator), and doesn't need to know about candidates and its index into that collection. I would prefer to maintain the separation of concerns.

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -8366,8 +8375,9 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
 
     // refresh the candidates map
     candidates.erase(candidates.begin());
-    if (! tracker.advance().at_end()) {
-      candidates[tracker.entry_name()] = tracker_idx;
+    tracker.advance();
+    if (tracker.next_candidate(candidates, tracker_idx)) {
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc]: As a result, this is a conditional with side-effects.

Comment (ivancich) [reply src/rgw/rgw_rados.cc]: Or I should have said, it's an empty conditional where the effect is all due to side-effects.

**Comment by cbodley** [comment]
@ivancich i moved next_candidate() into a lambda that returns void

**Comment by cbodley** [comment]
passed qa http://pulpito.ceph.com/cbodley-2020-02-28_20:21:50-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**